File: ecps_uv/observability/telemetry.py
"""
OpenTelemetry integration for ECPS.

This module provides telemetry integration using OpenTelemetry (OTLP),
including traces, metrics, and logs for comprehensive monitoring and debugging.
"""

import logging
import time
from typing import Any, Dict, List, Optional, Union

import uv

# OpenTelemetry imports
from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
from opentelemetry.context import Context, get_current

logger = logging.getLogger("ecps_uv.observability.telemetry")


class ECPSTelemetry:
    """
    Telemetry integration for ECPS using OpenTelemetry.
    
    This class provides a unified interface for collecting telemetry data
    (traces, metrics, logs) from all components within the ECPS ecosystem.
    """
    
    def __init__(
        self,
        otlp_endpoint: Optional[str] = None,
        service_name: str = "ecps-service",
        loop: Optional[Any] = None,
    ):
        """
        Initialize OpenTelemetry integration.
        
        Args:
            otlp_endpoint: OpenTelemetry endpoint URL (e.g., "http://localhost:4317")
            service_name: Name of the service for telemetry data
            loop: UV event loop to use
        """
        self.otlp_endpoint = otlp_endpoint or "http://localhost:4317"
        self.service_name = service_name
        self.loop = loop or uv.Loop.current()
        
        # TracerProvider for distributed tracing
        self.tracer_provider = None
        self.tracer = None
        
        # MeterProvider for metrics
        self.meter_provider = None
        self.meter = None
        
        # Initialize if endpoint is provided
        if otlp_endpoint:
            self._initialize_telemetry()
        else:
            logger.warning("No OTLP endpoint provided, telemetry will be disabled")
    
    def _initialize_telemetry(self):
        """Initialize OpenTelemetry providers and exporters."""
        # Initialize TracerProvider
        self.tracer_provider = TracerProvider()
        trace.set_tracer_provider(self.tracer_provider)
        
        # Create and register OTLP exporter for traces
        otlp_span_exporter = OTLPSpanExporter(endpoint=self.otlp_endpoint)
        span_processor = BatchSpanProcessor(otlp_span_exporter)
        self.tracer_provider.add_span_processor(span_processor)
        
        # Initialize MeterProvider
        metric_reader = PeriodicExportingMetricReader(
            exporter=OTLPMetricExporter(endpoint=self.otlp_endpoint),
            export_interval_millis=15000,  # Export every 15 seconds
        )
        self.meter_provider = MeterProvider(metric_readers=[metric_reader])
        metrics.set_meter_provider(self.meter_provider)
        
        # Create tracer and meter for this service
        self.tracer = trace.get_tracer(__name__, schemaUrl="https://ecps.io/schemas/1.0")
        self.meter = metrics.get_meter(__name__, schemaUrl="https://ecps.io/schemas/1.0")
        
        # Create standard metrics
        self._create_standard_metrics()
        
        logger.info(f"Initialized OpenTelemetry with endpoint {self.otlp_endpoint}")
    
    def _create_standard_metrics(self):
        """Create standard metrics for ECPS operations."""
        if not self.meter:
            return
        
        # EAP latency
        self.eap_latency = self.meter.create_histogram(
            name="eap.latency_ms",
            description="Latency of action execution in milliseconds",
            unit="ms",
        )
        
        # LTP frame size
        self.ltp_frame_bytes = self.meter.create_histogram(
            name="ltp.frame_bytes",
            description="Size of LTP frames in bytes",
            unit="bytes",
        )
        
        # MCP prompt size
        self.mcp_prompt_chars = self.meter.create_histogram(
            name="mcp.prompt_chars",
            description="Length of MCP prompts in characters",
            unit="chars",
        )
        
        # MEP query latency
        self.mep_query_latency = self.meter.create_histogram(
            name="mep.query_latency_ms",
            description="Latency of memory queries in milliseconds",
            unit="ms",
        )
        
        # Transport bytes
        self.transport_bytes_sent = self.meter.create_counter(
            name="transport.bytes_sent",
            description="Number of bytes sent over transport",
            unit="bytes",
        )
        
        self.transport_bytes_received = self.meter.create_counter(
            name="transport.bytes_received",
            description="Number of bytes received over transport",
            unit="bytes",
        )
    
    def create_span(
        self,
        name: str,
        context: Optional[Context] = None,
        kind: Optional[trace.SpanKind] = None,
        attributes: Optional[Dict[str, Any]] = None,
    ) -> trace.Span:
        """
        Create a new span for tracing.
        
        Args:
            name: Name of the span
            context: Parent context (optional)
            kind: Span kind (optional)
            attributes: Span attributes (optional)
            
        Returns:
            span: New span
        """
        if not self.tracer:
            # Return a no-op span if telemetry is disabled
            return trace.NoOpTracer().start_span(name)
        
        # Create span with context
        return self.tracer.start_span(
            name,
            context=context or get_current(),
            kind=kind or trace.SpanKind.INTERNAL,
            attributes=attributes or {},
        )
    
    def record_eap_latency(self, latency_ms: float, attributes: Optional[Dict[str, Any]] = None):
        """
        Record EAP action execution latency.
        
        Args:
            latency_ms: Latency in milliseconds
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        self.eap_latency.record(latency_ms, attributes=attributes)
    
    def record_ltp_frame_size(self, size_bytes: int, attributes: Optional[Dict[str, Any]] = None):
        """
        Record LTP frame size.
        
        Args:
            size_bytes: Size in bytes
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        self.ltp_frame_bytes.record(size_bytes, attributes=attributes)
    
    def record_mcp_prompt_size(self, chars: int, attributes: Optional[Dict[str, Any]] = None):
        """
        Record MCP prompt size.
        
        Args:
            chars: Number of characters
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        self.mcp_prompt_chars.record(chars, attributes=attributes)
    
    def record_mep_query_latency(self, latency_ms: float, attributes: Optional[Dict[str, Any]] = None):
        """
        Record MEP query latency.
        
        Args:
            latency_ms: Latency in milliseconds
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        self.mep_query_latency.record(latency_ms, attributes=attributes)
    
    def record_bytes_sent(self, bytes_count: int, transport_type: str, attributes: Optional[Dict[str, Any]] = None):
        """
        Record bytes sent over transport.
        
        Args:
            bytes_count: Number of bytes
            transport_type: Type of transport (e.g., "dds", "grpc", "mqtt")
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        all_attributes = {"transport_type": transport_type}
        if attributes:
            all_attributes.update(attributes)
        
        self.transport_bytes_sent.add(bytes_count, attributes=all_attributes)
    
    def record_bytes_received(self, bytes_count: int, transport_type: str, attributes: Optional[Dict[str, Any]] = None):
        """
        Record bytes received over transport.
        
        Args:
            bytes_count: Number of bytes
            transport_type: Type of transport (e.g., "dds", "grpc", "mqtt")
            attributes: Additional attributes for the metric
        """
        if not self.meter:
            return
        
        all_attributes = {"transport_type": transport_type}
        if attributes:
            all_attributes.update(attributes)
        
        self.transport_bytes_received.add(bytes_count, attributes=all_attributes)
    
    def inject_context_into_headers(self, context: Optional[Context], headers: Dict[str, str]) -> Dict[str, str]:
        """
        Inject trace context into headers for propagation.
        
        Args:
            context: Trace context to inject
            headers: HTTP headers to inject into
            
        Returns:
            headers: Updated HTTP headers with trace context
        """
        # Create new headers dictionary if needed
        headers = headers.copy() if headers else {}
        
        # Get the current context if none provided
        context = context or get_current()
        
        # Inject context into headers
        TraceContextTextMapPropagator().inject(headers, context)
        
        return headers
    
    def extract_context_from_headers(self, headers: Dict[str, str]) -> Context:
        """
        Extract trace context from headers.
        
        Args:
            headers: HTTP headers containing trace context
            
        Returns:
            context: Extracted trace context
        """
        # Extract context from headers
        return TraceContextTextMapPropagator().extract(headers)
    
    def with_span(self, name: str, kind: Optional[trace.SpanKind] = None, attributes: Optional[Dict[str, Any]] = None):
        """
        Decorator for tracing functions with spans.
        
        Args:
            name: Name of the span
            kind: Span kind (optional)
            attributes: Span attributes (optional)
            
        Returns:
            decorator: Function decorator
        """
        def decorator(func):
            async def wrapper(*args, **kwargs):
                with self.create_span(name, kind=kind, attributes=attributes) as span:
                    try:
                        result = await func(*args, **kwargs)
                        return result
                    except Exception as e:
                        span.record_exception(e)
                        raise
            
            return wrapper
        
        return decorator
    
    async def shutdown(self):
        """Shut down telemetry providers and exporters."""
        if self.tracer_provider:
            await self.tracer_provider.shutdown()
        
        if self.meter_provider:
            await self.meter_provider.shutdown()
        
        logger.info("Telemetry providers shut down")

File: ecps-go/pkg/observability/telemetry.go
// Package observability provides telemetry functionality for ECPS.
package observability

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/metric/metricdata"
	metricexport "go.opentelemetry.io/otel/sdk/metric/metricexporter"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/ecps/ecps-go/pkg/core"
)

// ECPSTelemetry implements the Telemetry interface for OpenTelemetry.
type ECPSTelemetry struct {
	tracer          trace.Tracer
	tracerProvider  *sdktrace.TracerProvider
	meterProvider   metric.MeterProvider
	meter           metric.Meter
	propagator      propagation.TextMapPropagator
	logger          core.Logger
	serviceName     string
	otlpEndpoint    string
	
	// Metrics
	eapLatency         metric.Float64Histogram
	ltpFrameBytes      metric.Int64Histogram
	mcpPromptChars     metric.Int64Histogram
	mepQueryLatency    metric.Float64Histogram
	transportBytesSent metric.Int64Counter
	transportBytesRecv metric.Int64Counter
}

// NewTelemetry creates a new ECPSTelemetry instance.
func NewTelemetry(
	otlpEndpoint string,
	serviceName string,
	logger core.Logger,
) (*ECPSTelemetry, error) {
	if logger == nil {
		logger = core.NewDefaultLogger()
	}
	
	t := &ECPSTelemetry{
		otlpEndpoint: otlpEndpoint,
		serviceName:  serviceName,
		logger:       logger,
	}
	
	// Initialize OpenTelemetry
	if err := t.initOpenTelemetry(); err != nil {
		return nil, fmt.Errorf("failed to initialize OpenTelemetry: %w", err)
	}
	
	return t, nil
}

// initOpenTelemetry initializes OpenTelemetry components.
func (t *ECPSTelemetry) initOpenTelemetry() error {
	// Create resource
	res, err := resource.New(context.Background(),
		resource.WithAttributes(
			semconv.ServiceNameKey.String(t.serviceName),
			semconv.ServiceVersionKey.String("1.0.0"),
		),
	)
	if err != nil {
		return fmt.Errorf("failed to create resource: %w", err)
	}

	// Set up OTLP exporter
	ctx := context.Background()
	conn, err := grpc.DialContext(ctx, t.otlpEndpoint, 
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithBlock(),
	)
	if err != nil {
		return fmt.Errorf("failed to create gRPC connection to OTLP endpoint: %w", err)
	}

	traceExporter, err := otlptrace.New(ctx, otlptracegrpc.NewClient(
		otlptracegrpc.WithGRPCConn(conn),
	))
	if err != nil {
		return fmt.Errorf("failed to create trace exporter: %w", err)
	}

	// Create trace provider
	bsp := sdktrace.NewBatchSpanProcessor(traceExporter)
	tracerProvider := sdktrace.NewTracerProvider(
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
		sdktrace.WithResource(res),
		sdktrace.WithSpanProcessor(bsp),
	)
	otel.SetTracerProvider(tracerProvider)
	t.tracerProvider = tracerProvider

	// Set up propagator
	propagator := propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	)
	otel.SetTextMapPropagator(propagator)
	t.propagator = propagator

	// Create tracer
	t.tracer = tracerProvider.Tracer(
		"github.com/ecps/ecps-go",
		trace.WithInstrumentationVersion("1.0.0"),
		trace.WithSchemaURL("https://ecps.io/schemas/1.0"),
	)

	// Create meter provider
	meterProvider := otel.GetMeterProvider()
	t.meterProvider = meterProvider

	// Create meter
	meter := meterProvider.Meter(
		"github.com/ecps/ecps-go",
		metric.WithInstrumentationVersion("1.0.0"),
		metric.WithSchemaURL("https://ecps.io/schemas/1.0"),
	)
	t.meter = meter

	// Create standard metrics
	return t.createStandardMetrics()
}

// createStandardMetrics creates standard metrics for ECPS operations.
func (t *ECPSTelemetry) createStandardMetrics() error {
	var err error

	// EAP latency
	t.eapLatency, err = t.meter.Float64Histogram(
		"eap.latency_ms",
		metric.WithDescription("Latency of action execution in milliseconds"),
		metric.WithUnit("ms"),
	)
	if err != nil {
		return fmt.Errorf("failed to create eap.latency_ms: %w", err)
	}

	// LTP frame size
	t.ltpFrameBytes, err = t.meter.Int64Histogram(
		"ltp.frame_bytes",
		metric.WithDescription("Size of LTP frames in bytes"),
		metric.WithUnit("bytes"),
	)
	if err != nil {
		return fmt.Errorf("failed to create ltp.frame_bytes: %w", err)
	}

	// MCP prompt size
	t.mcpPromptChars, err = t.meter.Int64Histogram(
		"mcp.prompt_chars",
		metric.WithDescription("Length of MCP prompts in characters"),
		metric.WithUnit("chars"),
	)
	if err != nil {
		return fmt.Errorf("failed to create mcp.prompt_chars: %w", err)
	}

	// MEP query latency
	t.mepQueryLatency, err = t.meter.Float64Histogram(
		"mep.query_latency_ms",
		metric.WithDescription("Latency of memory queries in milliseconds"),
		metric.WithUnit("ms"),
	)
	if err != nil {
		return fmt.Errorf("failed to create mep.query_latency_ms: %w", err)
	}

	// Transport bytes sent
	t.transportBytesSent, err = t.meter.Int64Counter(
		"transport.bytes_sent",
		metric.WithDescription("Number of bytes sent over transport"),
		metric.WithUnit("bytes"),
	)
	if err != nil {
		return fmt.Errorf("failed to create transport.bytes_sent: %w", err)
	}

	// Transport bytes received
	t.transportBytesRecv, err = t.meter.Int64Counter(
		"transport.bytes_received",
		metric.WithDescription("Number of bytes received over transport"),
		metric.WithUnit("bytes"),
	)
	if err != nil {
		return fmt.Errorf("failed to create transport.bytes_received: %w", err)
	}

	return nil
}

// Tracer returns the OpenTelemetry tracer.
func (t *ECPSTelemetry) Tracer() trace.Tracer {
	return t.tracer
}

// CreateSpan creates a new span for tracing.
func (t *ECPSTelemetry) CreateSpan(
	ctx context.Context,
	name string,
	kind trace.SpanKind,
	attrs map[string]interface{},
) (context.Context, trace.Span) {
	// Convert attributes
	attributes := make([]attribute.KeyValue, 0, len(attrs))
	for k, v := range attrs {
		attributes = append(attributes, attribute.String(k, fmt.Sprintf("%v", v)))
	}

	// Start span
	ctx, span := t.tracer.Start(
		ctx,
		name,
		trace.WithSpanKind(kind),
		trace.WithAttributes(attributes...),
	)

	return ctx, span
}

// RecordEAPLatency records EAP action execution latency.
func (t *ECPSTelemetry) RecordEAPLatency(latencyMs float64, attrs map[string]string) {
	if t.eapLatency != nil {
		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.eapLatency.Record(context.Background(), latencyMs, metric.WithAttributes(attributes...))
	}
}

// RecordLTPFrameSize records LTP frame size.
func (t *ECPSTelemetry) RecordLTPFrameSize(sizeBytes int, attrs map[string]string) {
	if t.ltpFrameBytes != nil {
		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.ltpFrameBytes.Record(context.Background(), int64(sizeBytes), metric.WithAttributes(attributes...))
	}
}

// RecordMCPPromptSize records MCP prompt size.
func (t *ECPSTelemetry) RecordMCPPromptSize(chars int, attrs map[string]string) {
	if t.mcpPromptChars != nil {
		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.mcpPromptChars.Record(context.Background(), int64(chars), metric.WithAttributes(attributes...))
	}
}

// RecordMEPQueryLatency records MEP query latency.
func (t *ECPSTelemetry) RecordMEPQueryLatency(latencyMs float64, attrs map[string]string) {
	if t.mepQueryLatency != nil {
		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.mepQueryLatency.Record(context.Background(), latencyMs, metric.WithAttributes(attributes...))
	}
}

// RecordBytesSent records bytes sent over transport.
func (t *ECPSTelemetry) RecordBytesSent(bytesCount int, transportType string, attrs map[string]string) {
	if t.transportBytesSent != nil {
		// Add transport type to attributes
		if attrs == nil {
			attrs = make(map[string]string)
		}
		attrs["transport_type"] = transportType

		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.transportBytesSent.Add(context.Background(), int64(bytesCount), metric.WithAttributes(attributes...))
	}
}

// RecordBytesReceived records bytes received over transport.
func (t *ECPSTelemetry) RecordBytesReceived(bytesCount int, transportType string, attrs map[string]string) {
	if t.transportBytesRecv != nil {
		// Add transport type to attributes
		if attrs == nil {
			attrs = make(map[string]string)
		}
		attrs["transport_type"] = transportType

		// Convert attributes
		attributes := make([]attribute.KeyValue, 0, len(attrs))
		for k, v := range attrs {
			attributes = append(attributes, attribute.String(k, v))
		}

		t.transportBytesRecv.Add(context.Background(), int64(bytesCount), metric.WithAttributes(attributes...))
	}
}

// InjectContext injects trace context into headers.
func (t *ECPSTelemetry) InjectContext(ctx context.Context, headers map[string]string) map[string]string {
	if headers == nil {
		headers = make(map[string]string)
	}

	// Create a carrier to inject context into
	carrier := propagation.MapCarrier{}
	t.propagator.Inject(ctx, carrier)

	// Copy to headers
	for k, v := range carrier {
		headers[k] = v
	}

	return headers
}

// ExtractContext extracts trace context from headers.
func (t *ECPSTelemetry) ExtractContext(ctx context.Context, headers map[string]string) context.Context {
	// Create a carrier from headers
	carrier := propagation.MapCarrier{}
	for k, v := range headers {
		carrier[k] = v
	}

	// Extract context
	return t.propagator.Extract(ctx, carrier)
}

// Shutdown shuts down telemetry providers and exporters.
func (t *ECPSTelemetry) Shutdown(ctx context.Context) error {
	// Shutdown tracer provider
	if t.tracerProvider != nil {
		if err := t.tracerProvider.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown tracer provider: %w", err)
		}
	}

	return nil
}

// WithSpan runs a function within a new span.
func (t *ECPSTelemetry) WithSpan(
	ctx context.Context,
	name string,
	kind trace.SpanKind,
	attrs map[string]interface{},
	fn func(context.Context) error,
) error {
	ctx, span := t.CreateSpan(ctx, name, kind, attrs)
	defer span.End()

	err := fn(ctx)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(trace.StatusCodeError, err.Error())
	}

	return err
}

// FormatAttributes formats attributes for logging.
func (t *ECPSTelemetry) FormatAttributes(attrs map[string]interface{}) string {
	if len(attrs) == 0 {
		return ""
	}

	result := "{"
	first := true
	for k, v := range attrs {
		if !first {
			result += ", "
		}
		result += fmt.Sprintf("%s=%v", k, v)
		first = false
	}
	result += "}"

	return result
}